File: app/src/debug/java/com/isardomains/sameview/ui/camera/PhotoPickerMimicContentProvider.java
package com.isardomains.sameview.ui.camera;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;
import android.provider.OpenableColumns;

import java.io.File;
import java.io.FileNotFoundException;

/**
 * Test-only ContentProvider simulating the Photo Picker / MediaStore split-source behaviour
 * that causes the double-rotation format-mismatch bug.
 *
 * This is a pure-Java class (no Kotlin stdlib) so it can run in the test APK's own process
 * without depending on the app APK's Kotlin runtime. android:exported="true" in the test
 * manifest is required so the app process (different UID from the test APK package) can
 * reach this provider via ContentResolver.
 *
 * Behaviour:
 * - URI without "require_original" query param → serves the pre-rotated file
 * - URI with "require_original=1" (appended by MediaStore.setRequireOriginal) → serves original
 */
public class PhotoPickerMimicContentProvider extends ContentProvider {

    public static final String AUTHORITY = "com.isardomains.sameview.test.photopickermimicmedia";
    private static final String PARAM_ORIGINAL_PATH = "original_path";
    private static final String PARAM_PREROTATED_PATH = "prerotated_path";

    @Override
    public boolean onCreate() {
        return true;
    }

    @Override
    public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundException {
        boolean useOriginal = uri.getQueryParameter("require_original") != null;
        String path = useOriginal
                ? uri.getQueryParameter(PARAM_ORIGINAL_PATH)
                : uri.getQueryParameter(PARAM_PREROTATED_PATH);
        if (path == null) {
            throw new FileNotFoundException("Missing path parameter in URI: " + uri);
        }
        File file = new File(path);
        if (!file.exists()) {
            throw new FileNotFoundException("File not found: " + path);
        }
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY);
    }

    @Override
    public String getType(Uri uri) {
        return "image/jpeg";
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection,
            String[] selectionArgs, String sortOrder) {
        String path = uri.getQueryParameter(PARAM_ORIGINAL_PATH);
        if (path == null) return null;
        File file = new File(path);
        MatrixCursor cursor = new MatrixCursor(
                new String[]{OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE});
        cursor.addRow(new Object[]{file.getName(), file.length()});
        return cursor;
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) { return null; }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) { return 0; }

    @Override
    public int update(Uri uri, ContentValues values, String selection,
            String[] selectionArgs) { return 0; }

    public static Uri uriFor(File originalFile, File prerotatedFile) {
        // require_original=1 is pre-embedded so the provider always serves the original file.
        // MediaStore.setRequireOriginal() on Android 16+ rejects non-MediaStore authorities
        // and throws IllegalArgumentException; resolveSourceUri() catches that and returns the
        // URI unchanged — so the flag must already be present for openFile() to see it.
        return new Uri.Builder()
                .scheme("content")
                .authority(AUTHORITY)
                .path("/image")
                .appendQueryParameter(PARAM_ORIGINAL_PATH, originalFile.getAbsolutePath())
                .appendQueryParameter(PARAM_PREROTATED_PATH, prerotatedFile.getAbsolutePath())
                .appendQueryParameter("require_original", "1")
                .build();
    }
}

File: app/src/androidTest/java/com/isardomains/sameview/ui/camera/PhotoPickerMimicContentProvider.kt
package com.isardomains.sameview.ui.camera

import android.content.ContentProvider
import android.content.ContentValues
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.provider.OpenableColumns
import java.io.File
import java.io.FileNotFoundException

/**
 * Test-only ContentProvider simulating the Photo Picker / MediaStore split-source behaviour
 * that causes the double-rotation format-mismatch bug.
 *
 * Real-device scenario reproduced here:
 * - A portrait HEIC is physically stored with landscape sensor pixels (e.g. rawW=4032, rawH=3024)
 *   plus EXIF ORIENTATION_ROTATE_90.
 * - MediaStore.openInputStream(uri) may return a transcoded pre-rotated JPEG with portrait pixels
 *   (rawW=3024, rawH=4032) and no EXIF rotation tag.
 * - MediaStore.openFileDescriptor(setRequireOriginal(uri)) returns the original HEIC with
 *   landscape pixels and EXIF ROTATE_90.
 *
 * Without the fix the reader combined portrait dims from the plain stream with ROTATE_90 from
 * the original, doubling the rotation to an incorrect landscape oriented result.
 *
 * This provider simulates the split by serving different files based on the presence of the
 * "require_original" query parameter that [android.provider.MediaStore.setRequireOriginal] appends:
 * - URI without "require_original" → serves the pre-rotated file (simulates transcoded stream)
 * - URI with "require_original=1" → serves the original file (simulates setRequireOriginal path)
 *
 * The provider authority is [AUTHORITY], which is NOT "media". Tests pass [AUTHORITY] explicitly
 * via the [com.isardomains.sameview.ui.camera.ReferenceImageMetadataReader.read]
 * `requireOriginalAuthorities` parameter so the reader applies setRequireOriginal to this URI,
 * exercising the production code path without touching real MediaStore.
 */
internal class PhotoPickerMimicContentProvider : ContentProvider() {

    override fun onCreate(): Boolean = true

    override fun openFile(uri: Uri, mode: String): ParcelFileDescriptor {
        val useOriginal = uri.getQueryParameter("require_original") != null
        val path = if (useOriginal) {
            uri.getQueryParameter(PARAM_ORIGINAL_PATH)
        } else {
            uri.getQueryParameter(PARAM_PREROTATED_PATH)
        } ?: throw FileNotFoundException("Missing path parameter in URI: $uri")
        val file = File(path)
        if (!file.exists()) throw FileNotFoundException("File not found: $path")
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
    }

    override fun getType(uri: Uri): String = "image/jpeg"

    override fun query(
        uri: Uri, projection: Array<String>?, selection: String?,
        selectionArgs: Array<String>?, sortOrder: String?
    ): Cursor? {
        val path = uri.getQueryParameter(PARAM_ORIGINAL_PATH) ?: return null
        val file = File(path)
        val cols = arrayOf(OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE)
        val cursor = MatrixCursor(cols)
        cursor.addRow(arrayOf(file.name, file.length()))
        return cursor
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? = null
    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int = 0
    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int = 0

    companion object {
        const val AUTHORITY = "com.isardomains.sameview.test.photopickermimicmedia"
        private const val PARAM_ORIGINAL_PATH = "original_path"
        private const val PARAM_PREROTATED_PATH = "prerotated_path"

        /**
         * Builds a URI that [PhotoPickerMimicContentProvider] can serve.
         *
         * @param originalFile The file with the original sensor-native pixels and EXIF orientation.
         * @param prerotatedFile The file with rotation baked into pixels and no EXIF rotation.
         */
        fun uriFor(originalFile: File, prerotatedFile: File): Uri = Uri.Builder()
            .scheme("content")
            .authority(AUTHORITY)
            .path("/image")
            .appendQueryParameter(PARAM_ORIGINAL_PATH, originalFile.absolutePath)
            .appendQueryParameter(PARAM_PREROTATED_PATH, prerotatedFile.absolutePath)
            .build()
    }
}
